### PR TITLE
miniconda

### PR DESCRIPTION
**Pull request by mloubout**
back to latest version, version that my hard coding was skiping is not there anymore so we shouldn't need that old version anymore

### CODE REVIEW

File: Dockerfile.pipelines [hunk 1]
```diff
@@ -93,10 +98,10 @@ RUN chown -R devito /home/devito
 USER devito
 WORKDIR /home/devito
 
-RUN if [ $installWithPip == "true" ] ; then \ 
-      python setup.py install --user ; \
+RUN if [ $installWithPip == "true" ] ; then \
+      python3 setup.py install --user ; \
       if [ $MPI_INSTALL == '1' ] ; then \
-        pip install --user -e .[extras] ; \
+        pip3 install --user -e .[extras] ; \
```
Comment (tmbgreaves) [inline Dockerfile.pipelines]:
```suggestion
        python3 -m pip install --user -e .[extras] ; \
```

This seems to be the way the world is going..

**Comment by tmbgreaves** [comment]
Versions on Travis using pip setup appear to be falling over with a pytest version issue -- looking for >3.6 but finding version 3.3. I guess 3.3 is the version preinstalled in the Travis environment and needs upgrading?

**Comment by tmbgreaves** [comment]
The package problem on the pip installer is coming from our base image -- we're building on Ubuntu Xenial which only has python 3.5, and we're requesting 3.6. The fix would be to either regress to 3.5, or to bump the base image up to Bionic which has 3.6 and 3.7 available. Moving to bionic sounds worthwhile anyway, as Xenial is pretty old now, but would need a few more changes to the Dockerfile (which I can easily do if that would be useful). Let me know which way you'd prefer to go on this :-)

File: .travis.yml [hunk 2]
```diff
@@ -60,37 +70,27 @@ matrix:
...
-  - conda info -a
   - if [[ "$MPI_INSTALL" == '1' ]]; then
       sudo apt-get install -y -q mpich libmpich-dev;
       export MPICH_CC=$DEVITO_ARCH;
     fi
+  # Setup anaconda
+  - if [[ $INSTALL_TYPE == 'conda' ]]; then
+      wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh;
+      bash miniconda.sh -b -p $HOME/miniconda;
+      export PATH="$HOME/miniconda/bin:$PATH";
```
Comment (tmbgreaves) [inline .travis.yml]:
```suggestion
      export PATH="/usr/local/miniconda/bin:$PATH";
```
Path fix for different install location

File: Dockerfile.pipelines [hunk 3]
```diff
@@ -78,11 +78,16 @@ RUN if [ "$MPI_INSTALL" == "1" ] ; then \
     fi
 
 # Set up conda
-RUN wget --no-verbose https://repo.continuum.io/miniconda/Miniconda3-4.5.4-Linux-x86_64.sh -O /tmp/miniconda.sh
-RUN bash /tmp/miniconda.sh -b -p /usr/local/miniconda
+RUN if [ $installWithPip == "true" ] ; then \ 
+      apt-get install -y python3.6-dev python3-pip ; \
+    else \
+      wget --no-verbose https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O /tmp/miniconda.sh ; \
+      bash /tmp/miniconda.sh -b -p /usr/local/miniconda ; \
```
Comment (tmbgreaves) [inline Dockerfile.pipelines]:
```suggestion
      bash /tmp/miniconda.sh -b -p /usr/local/miniconda ; \
      export PATH=/usr/local/miniconda/bin:$PATH
```
We need an extra bit of path setting here so the following conda commands work. It should be safe to leave the ENV command in line 90 as if there's no conda installed it won't have any effect.

Comment (mloubout) [reply Dockerfile.pipelines]: Yeah these lines are giving me trouble so far no luck having it to run. Like in travis I'm just trying to skip the conda installation to make sure we don't get lucky with pip getting so extras from conda.

**Comment by mloubout** [comment]
>directly calling pip to calling 'python3 -m pip
Yeah it's usually more stable and safe

>bump the base image up to Bionic
I'm fine with that

**Comment by mloubout** [comment]
"Beginning with Matplotlib 3.1, Python 3.6 or above is required."

Gotta have separate requirements now....

**Comment by tmbgreaves** [comment]
Testing with bionic, we hit the problem that ubuntu-toolchain-r only builds gcc9, so we'd be limited to 7 and 9 to test on bionic. Does that change the decision to move?

**Comment by mloubout** [comment]
Other are tested on travis s I'd be fine with it. @ggorman @FabioLuporini what do you think?

Also I kinda want to drop python 3.5 because some (lots?) dependencies don't support it anyway.

**Comment by tmbgreaves** [comment]
OK! I've pushed some changes as suggestions :-)

File: Dockerfile.pipelines [hunk 4]
```diff
@@ -93,10 +76,10 @@ RUN chown -R devito /home/devito
 USER devito
 WORKDIR /home/devito
 
-RUN if [ $installWithPip == "true" ] ; then \ 
-      python setup.py install --user ; \
+RUN if [ $installWithPip == "true" ] ; then \
+      python3 setup.py install --user ; \
```
Comment (FabioLuporini) [inline Dockerfile.pipelines]: why isn't ``python`` the default ``python3``?

Comment (mloubout) [reply Dockerfile.pipelines]: default is not the 3.6 which is called python3

**Comment by mloubout** [comment]
SO mainly all green except:
- waiting on inducer to burn new realease of codepy
- One failure with conda and openmp8 for some reason I don't understand

File: .travis.yml [hunk 5]
```diff
@@ -20,8 +20,20 @@ matrix:
             - gcc-4.9
             - g++-4.9
       env: DEVITO_ARCH=gcc-4.9 DEVITO_OPENMP=0 INSTALL_TYPE=pip_setup RUN_EXAMPLES=False MPI_INSTALL=1
-    - os: linux
-      python: "3.6"
+    - python: 3.7
```
Comment (FabioLuporini) [inline .travis.yml]: is this a new build??

File: .travis.yml [hunk 6]
```diff
@@ -60,37 +75,27 @@ matrix:
             - gcc-7
             - g++-7
       env: DEVITO_ARCH=gcc-7 DEVITO_OPENMP=0 DEVITO_BACKEND=ops YC_CXX=g++-7 INSTALL_TYPE=conda RUN_EXAMPLES=False MPI_INSTALL=0
-  allow_failures:
```
Comment (FabioLuporini) [inline .travis.yml]: what was this doing before? I have absolutely no idea

File: .travis.yml [hunk 7]
```diff
@@ -30,8 +42,9 @@ matrix:
             - gcc-5
             - g++-5
       env: DEVITO_ARCH=gcc-5 DEVITO_OPENMP=0 RUN_EXAMPLES=False INSTALL_TYPE=conda MPI_INSTALL=0
-    - os: linux
-      python: "3.6"
+    - python: 3.7
+      dist: xenial
```
Comment (FabioLuporini) [inline .travis.yml]: why fixing xenial?

File: .travis.yml [hunk 8]
```diff
@@ -60,37 +75,27 @@ matrix:
...
-      env: DEVITO_ARCH=clang DEVITO_OPENMP=0
 
 addons:
   apt:
     sources:
       - ubuntu-toolchain-r-test     # For gcc 4.9, 5 and 7
 
 before_install:
-  # Setup anaconda
-  - if [[ "$TRAVIS_PYTHON_VERSION" == "2.7" ]]; then
```
Comment (FabioLuporini) [inline .travis.yml]: here you're just dropping the hack documented in the issues page right? ie pulling an old miniconda version for reasons I now don't remember

File: .travis.yml [hunk 9]
```diff
@@ -60,37 +75,27 @@ matrix:
...
-  - hash -r
-  - conda config --set always_yes yes --set changeps1 no
-  - conda update -q conda
-  # Useful for debugging any issues with conda
-  - conda info -a
   - if [[ "$MPI_INSTALL" == '1' ]]; then
       sudo apt-get install -y -q mpich libmpich-dev;
       export MPICH_CC=$DEVITO_ARCH;
     fi
+  # Setup anaconda
```
Comment (FabioLuporini) [inline .travis.yml]: just wondering why has this been moved after the ``if MPI_INSTALL`` block?

File: Dockerfile.pipelines [hunk 10]
```diff
@@ -1,15 +1,8 @@
-FROM ubuntu:xenial
+FROM ubuntu:bionic
```
Comment (FabioLuporini) [inline Dockerfile.pipelines]: why this change?

File: Dockerfile.pipelines [hunk 11]
```diff
@@ -1,15 +1,8 @@
...
 
 # This DockerFile is looked after by
 MAINTAINER Tim Greaves <tim.greaves@imperial.ac.uk>
 
-# Add the ubuntu-toolchain-r test ppa
```
Comment (FabioLuporini) [inline Dockerfile.pipelines]: is this a consequence of not using xenial?

File: Dockerfile.pipelines [hunk 12]
```diff
@@ -25,11 +18,9 @@ RUN apt-get -y dist-upgrade
 # Needed for the conda and devito installs later
 RUN apt-get -y install wget bzip2 git make
 
-# Default gcc version to install
-ARG gccVersion=8
-ENV DEVITO_ARCH=gcc-$gccVersion
-# Set Yask variables in case they are needed
-ENV YC_CXX=g++-$gccVersion
+# Set environment
```
Comment (FabioLuporini) [inline Dockerfile.pipelines]: why?

File: Dockerfile.pipelines [hunk 13]
```diff
@@ -25,11 +18,9 @@ RUN apt-get -y dist-upgrade
 # Needed for the conda and devito installs later
 RUN apt-get -y install wget bzip2 git make
 
-# Default gcc version to install
-ARG gccVersion=8
-ENV DEVITO_ARCH=gcc-$gccVersion
```
Comment (FabioLuporini) [inline Dockerfile.pipelines]: I'm concerned by this change. This is not parametric anymore? why should it be the case?

File: setup.py [hunk 14]
```diff
@@ -19,12 +19,21 @@
 
 opt_reqs = []
 opt_links = []
+dict_opt_reqs = {}
```
Comment (FabioLuporini) [inline setup.py]: call it ``extras_require``?

File: .travis.yml (hunk 7 above)
Comment (mloubout) [reply .travis.yml]: 3.7 only supported on xenial dist, so has to be put there.

File: .travis.yml (hunk 6 above)
Comment (mloubout) [reply .travis.yml]: With was allowing failure on [ython 2.7 which is not tested pr supported so was useless

File: .travis.yml (hunk 5 above)
Comment (mloubout) [reply .travis.yml]: Yes, now two builds, one for 3.6 and one for 3.7

File: .travis.yml (hunk 8 above)
Comment (mloubout) [reply .travis.yml]: Yes. A previous version had issues with numpy so had to be skipped. Newest seem to have fixed it so we can get back to latest, and dropped untested 2.7

File: .travis.yml (hunk 9 above)
Comment (mloubout) [reply .travis.yml]: No special reason, was a tab issue in the yaml was moving stuff around to find it, I can change it back doesn't really matter what order

File: Dockerfile.pipelines (hunk 10 above)
Comment (mloubout) [reply Dockerfile.pipelines]: @tmbgreaves commented above about it, necessary for python versions

File: setup.py (hunk 14 above)
Comment (mloubout) [reply setup.py]: Can change name yes.

File: Dockerfile.pipelines (hunk 13 above)
Comment (mloubout) [reply Dockerfile.pipelines]: Not sure about these changes, was necessary for bionic.

File: .travis.yml (hunk 5 above)
Comment (FabioLuporini) [reply .travis.yml]: I don't think a whole new build is a good idea, as this will increase the turnaround time by at least another hour (maybe two). Can we not use 3.6 in one of the existing builds?

File: .travis.yml (hunk 8 above)
Comment (FabioLuporini) [reply .travis.yml]: OK

File: .travis.yml (hunk 9 above)
Comment (FabioLuporini) [reply .travis.yml]: it would only make the diff clearer, but otherwise no significant reason (had to compare the lines one by one)

File: Dockerfile.pipelines (hunk 13 above)
Comment (FabioLuporini) [reply Dockerfile.pipelines]: I think we then have to step back -- we absolutely need to test multiple gcc versions, in particular the newest ones

Comment (mloubout) [reply Dockerfile.pipelines]: We still do, it's just that now gcc-8 comes default so no need to add extra install, so it's the exact same setup as before just "easier" to install with defaults in bionic

Comment (tmbgreaves) [reply Dockerfile.pipelines]: I think I need to regress some changes here, actually -- I was assuming (in error) something special about toolchain-r which wasn't provided by system packages.

File: Dockerfile.pipelines (hunk 11 above)
Comment (tmbgreaves) [reply Dockerfile.pipelines]: Yes - toolchain-r drops all except 9 when you move to xenial. But you do get 9 - is this useful to test against?

Comment (tmbgreaves) [reply Dockerfile.pipelines]: Dropped this change - see thoughts below on gcc versioning.

**Comment by tmbgreaves** [comment]
Thoughts on gcc:

Versions available from bionic are 5, 6, 7, and 8, plus 9 from toolchain-r. I've provisionally added a 9 builder -- change as required. 4 is no longer available on bionic so I've moved the pip builder to 8 -- again change as required.

**Comment by tmbgreaves** [comment]
Turns out the gcc installs from bionic aren't as straightforward as expected -- hacking on it now.

**Comment by tmbgreaves** [comment]
It appears that adding in ubuntu-toolchain-r breaks installs of the system compiler packages, so gcc 9 from there might be better to leave for now. If 9 would be useful, we could add in the toolchain-r ppa bit as a conditional on gcc 9 being requested.

**Comment by tmbgreaves** [comment]
OK -- I think latest commit has fixed things and this now has working 5, 7, and 8 from system packages.

File: .travis.yml (hunk 5 above)
Comment (mloubout) [reply .travis.yml]: Well no, this test python 3.6 and 3.7 with pip then there is the conda builds. Can try to merge one of the conda with one of the two pip but will reduce conda testing. I'm not sure about turnaround time as they supposed to run toghether so don't see why it would be slower

Comment (mloubout) [reply .travis.yml]: Removed one, added one to pipelines instead as it is faster

File: .travis.yml [hunk 15]
```diff
@@ -134,6 +128,8 @@ script:
   - if [[ $RUN_EXAMPLES == 'True' ]]; then python examples/misc/linalg.py mat-vec mat-mat-sum transpose-mat-vec; fi
   - if [[ $RUN_EXAMPLES == 'True' ]]; then python examples/seismic/tti/tti_example.py -a; fi
   - if [[ $RUN_EXAMPLES == 'True' ]]; then python examples/seismic/tti/tti_example.py -a --noazimuth; fi
+  - if [[ $RUN_EXAMPLES == 'True' ]]; then python examples/cfd/example_diffusion.py; fi
```
Comment (FabioLuporini) [inline .travis.yml]: why do we have to run it once with ``python`` and once with ``pytest``?

File: azure-pipelines.yml [hunk 16]
```diff
@@ -12,48 +12,62 @@ jobs:
     vmImage: 'Ubuntu 16.04'
   strategy:
     matrix:
-      pip-gcc4:
-        gccVersion: '4.9'
+      pip-gcc8-py36:
```
Comment (FabioLuporini) [inline azure-pipelines.yml]: this is called gcc8, but below it says gccVersions=6

I'm not sure I understand why we want to add a new build

File: .travis.yml [hunk 17]
```diff
@@ -10,18 +10,19 @@ env:
...
         apt:
           sources:
             - ubuntu-toolchain-r-test     # For gcc 4.9, 5 and 7
           packages:
-            - gcc-4.9
-            - g++-4.9
-      env: DEVITO_ARCH=gcc-4.9 DEVITO_OPENMP=0 INSTALL_TYPE=pip_setup RUN_EXAMPLES=False MPI_INSTALL=1
-    - os: linux
-      python: "3.6"
+            - gcc-7
```
Comment (FabioLuporini) [inline .travis.yml]: why do you wanna drop 4.9? it's still (unfortunately) one of the most used gcc versions, so it should be tested right?

what we may want is testing gcc9 ...

Comment (tmbgreaves) [reply .travis.yml]: 4.9 - Lack of a build for bionic, there only appears to be a stub set of packages on the system set, and none on ubuntu-toolchain-r, but I haven't looked ultra-carefully.

9 - I think this should work with some bash conditionals wrapped round the gcc install to only turn on the ubuntu-toolchain-r repo when installing 9.

I'm a bit tied up with things this week but will do my best to fit in having a look at it if nobody beats me to it.

File: azure-pipelines.yml (hunk 16 above)
Comment (mloubout) [reply azure-pipelines.yml]: We want to test both python 3.7 and 3.6 so that's two buildm then the conda ones with the examples and stuff

File: .travis.yml (hunk 15 above)
Comment (mloubout) [reply .travis.yml]: There is a test in it that used to be run in a weird way I just moved it here because it needs example install for matplotlib

File: azure-pipelines.yml (hunk 16 above)
Comment (mloubout) [reply azure-pipelines.yml]: Can have up to 10 at same time on pipeline so it' fine I think

**Comment by tmbgreaves** [comment]
A thought -- do we even need to touch system python here or could we do everything using conda? Ie, if we don't care about system python, it might be feasible to stick with xenial for the purposes of getting this PR through, and deal with the xenial->bionic transition separately. I think moving off xenial is a good thing to do at some point soon given we're now a year past the following LTS (bionic) release.

**Comment by mloubout** [comment]
Well system python is used to test the pip way. The only issue left currently is blosc failing to compile with gcc-9 and everything else should be fine. I added a line to force-link the installed gcc version and python so this part is fixed.

File: azure-pipelines.yml [hunk 18]
```diff
@@ -12,48 +12,54 @@ jobs:
     vmImage: 'Ubuntu 16.04'
   strategy:
     matrix:
-      pip-gcc4:
-        gccVersion: '4.9'
+      pip-gcc6-py36:
+        gccVersion: '6'
         DEVITO_OPENMP: '0'
-        MPI_INSTALL: '1'
+        MPI_INSTALL: '0'
```
Comment (FabioLuporini) [inline azure-pipelines.yml]: this should be 1?
